### PR TITLE
[KEP] Update KEP #993 - Add configuration to retry mechanism for ProvisioningRequests

### PR DESCRIPTION
**Pull request by PBundyra**
#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
Propose changes to introduce configuration to retry mechanism for ProvisioningRequests

#### Which issue(s) this PR fixes:

Part of #1353 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by PBundyra** [comment]
/assign @mimowo 
cc @mwielgus @yaroslava-serdiuk

**Comment by trasc** [comment]
In my opinion we should let individual admission check controllers implement their own retry policy, not all ACCs need this and having another actor managing its ACStates makes room for conflicts or needs additional logic to avoid them.

File: keps/993-two-phase-admission/README.md [hunk 1]
```diff
@@ -187,6 +188,29 @@ type AdmissionCheckParametersReference struct {
...
+	//
+	// Default values are:
+	//
+	// defaultRequeuingStrategyAC:
+	//	timestamp: Eviction
+	//	backoffLimitCount: 5
+	//	backoffBaseSeconds: 60
+	//	backoffMaxSeconds: 600
+	//
+	// +listType=map
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply keps/993-two-phase-admission/README.md]: Done

File: keps/993-two-phase-admission/README.md [hunk 2]
```diff
@@ -187,6 +188,29 @@ type AdmissionCheckParametersReference struct {
 	// Name is the name of the resource being referenced.
 	Name string `json:"name"`
 }
+
+// Configuration is the Schema for the kueueconfigurations API
+type Configuration struct {
+	...
+	// admissionCheckRetryStrategy defines strategy for retrying AdmissionChecks
+	// +optional
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply keps/993-two-phase-admission/README.md]: Done, thanks

File: keps/993-two-phase-admission/README.md [hunk 3]
```diff
@@ -187,6 +188,29 @@ type AdmissionCheckParametersReference struct {
...
+	// Default values are:
+	//
+	// defaultRequeuingStrategyAC:
+	//	timestamp: Eviction
+	//	backoffLimitCount: 5
+	//	backoffBaseSeconds: 60
+	//	backoffMaxSeconds: 600
+	//
+	// +listType=map
+	DefaultRequeuingStrategy *RequeuingStrategy `json:defaultRequeuingStrategy`
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply keps/993-two-phase-admission/README.md:210]: Yes, and it introduces a field `releaseQuota` that allows users to release quota if the ProvisioningRequest fails

**Comment by PBundyra** [comment]
> In my opinion we should let individual admission check controllers implement their own retry policy, not all ACCs need this and having another actor managing its ACStates makes room for conflicts or needs additional logic to avoid them.

Thanks for the review Traian. I considered it, and change the proposal to cover only ProvisioningRequests

File: keps/993-two-phase-admission/README.md [hunk 4]
```diff
@@ -187,6 +188,34 @@ type AdmissionCheckParametersReference struct {
...
+	//	backoffBaseSeconds: 60
+	//	backoffMaxSeconds: 1800
+	DefaultRequeuingStrategy *RequeuingStrategy `json:defaultRequeuingStrategy`
+
+	// releaseQuota defines whether the reserved quota should be released during retry.
+	// If the releaseQuota is set not false, controller will delete old ProvisioningRequest
+	// and create a new one without releasing quota.
+	//
+	// Defaults to true.
+	ReleaseQuota *bool `json:releaseQuota`
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply keps/993-two-phase-admission/README.md]: Fixed

File: keps/993-two-phase-admission/README.md [hunk 5]
```diff
@@ -232,8 +261,15 @@ are still valid (check quota/preemptions/reclamation) and admit the workload (do
 newly identified preemptions, if needed), setting the `Admitted`
 condition to True.
 
-If any check is switched to "False", with reason "Reject" or "Retry", the workload is either
-completely rejected or switched back to suspend state (with retry delay).
+If a ProvisioningRequest check is switched to `Retry` state, the Workload will be requeued
+or deactivated, with respect to the `ProvisioningRequestRetryStrategy.defaultRequeuingStrategy` defined in
+Kueue's global Configuration. If the Workload is requeued, controller should update Workload's `.status.requeueState`
+accordingly. If the limit of number of retries has been surpassed, the controller will mark the check as `Rejected`,
```
Comment (trasc) [inline keps/993-two-phase-admission/README.md:267]:
```suggestion
accordingly. If the limit of number of retries has been surpassed, the provisioning admission check controller will mark the check as `Rejected`,
```

File: keps/993-two-phase-admission/README.md [hunk 6]
```diff
@@ -187,6 +188,34 @@ type AdmissionCheckParametersReference struct {
...
+	// defaultRequeuingStrategyAC:
+	//	timestamp: Eviction
+	//	backoffLimitCount: 3
+	//	backoffBaseSeconds: 60
+	//	backoffMaxSeconds: 1800
+	DefaultRequeuingStrategy *RequeuingStrategy `json:defaultRequeuingStrategy`
+
+	// releaseQuota defines whether the reserved quota should be released during retry.
+	// If the releaseQuota is set not false, controller will delete old ProvisioningRequest
+	// and create a new one without releasing quota.
```
Comment (trasc) [inline keps/993-two-phase-admission/README.md:214]:
```suggestion
	// releaseQuota defines whether the reserved quota should be released during when its ProvisioningRequest  fails.
	// If the releaseQuota is set not false, controller will delete old ProvisioningRequest
	// and create a new one without setting the workload's adission check state to `Retry`.
```

If an admission check state is set to `Retry` the workload controller will release the quota.

File: keps/993-two-phase-admission/README.md [hunk 7]
```diff
@@ -232,8 +261,15 @@ are still valid (check quota/preemptions/reclamation) and admit the workload (do
 newly identified preemptions, if needed), setting the `Admitted`
 condition to True.
 
-If any check is switched to "False", with reason "Reject" or "Retry", the workload is either
-completely rejected or switched back to suspend state (with retry delay).
+If a ProvisioningRequest check is switched to `Retry` state, the Workload will be requeued
+or deactivated, with respect to the `ProvisioningRequestRetryStrategy.defaultRequeuingStrategy` defined in
+Kueue's global Configuration. If the Workload is requeued, controller should update Workload's `.status.requeueState`
```
Comment (trasc) [inline keps/993-two-phase-admission/README.md:266]: >  If the Workload is requeued, controller ...

Which Controller should do this? the workload controller or the admission check controller.

File: keps/993-two-phase-admission/README.md [hunk 8]
```diff
@@ -187,6 +188,34 @@ type AdmissionCheckParametersReference struct {
...
+	// The strategy is applied to all classes of ProvisioningRequest.
+	//
+	// Default values are:
+	//
+	// defaultRequeuingStrategyAC:
+	//	timestamp: Eviction
+	//	backoffLimitCount: 3
+	//	backoffBaseSeconds: 60
+	//	backoffMaxSeconds: 1800
+	DefaultRequeuingStrategy *RequeuingStrategy `json:defaultRequeuingStrategy`
```
Comment (yaroslava-serdiuk) [inline keps/993-two-phase-admission/README.md:210]: nit: Do we need default suffix in the name?

**Comment by PBundyra** [comment]
I took it a step further and moved the API to the ProvisioningRequestConfig API in #3278, so I'm closing this one.

**Pull request closed without merging**